File: finalyearproject/repos/dkd-platform-libs/sdk/typescript/src/dlq.ts
// HAND-AUTHORED platform primitive (NOT dkdgen-generated).
// Dead-letter quarantine with per-key park-and-freeze (SA-MSG-09/10). A poison money/custody/
// inventory event is parked to the DLQ and its aggregate_key is FROZEN — only that key stops
// advancing while every other key keeps flowing. Nothing is ever silently dropped.
// One of five byte-identical language implementations sharing the SAME contract (PL-02).
import type { DbExecutor } from "./outbox.js";

export interface ParkedEvent {
  eventId: string;
  topic: string;
  key: string;
  /** Serialized canonical JSON of the offending event — cast to jsonb at the boundary. */
  payload: string;
  error: string;
  /** The single aggregate key to freeze (e.g. WLT/ESC/TXN, PPID). */
  aggregateKey: string;
}

export class Dlq {
  /** Quarantine a poison event and freeze ONLY its aggregate key. */
  static async park(db: DbExecutor, ev: ParkedEvent): Promise<void> {
    await db.query(
      "INSERT INTO dlq(event_id, topic, key, payload, error, aggregate_key, parked_at) " +
        "VALUES ($1,$2,$3,$4::jsonb,$5,$6,now())",
      [ev.eventId, ev.topic, ev.key, ev.payload, ev.error, ev.aggregateKey],
    );
  }

  /**
   * True when this aggregate key has a parked (frozen) event. A consumer checks this before
   * processing to keep the frozen key stalled while other keys progress.
   */
  static async isKeyParked(db: DbExecutor, aggregateKey: string): Promise<boolean> {
    const res = await db.query(
      "SELECT 1 FROM dlq WHERE aggregate_key = $1 LIMIT 1",
      [aggregateKey],
    );
    return res.rowCount >= 1;
  }
}

File: finalyearproject/repos/dkd-platform-libs/sdk/go/dlq.go
package dkdplatform

import (
	"context"
	"fmt"
)

// DlqRecord is a poison event routed to the dead-letter store. AggregateKey is
// the ordering key that gets frozen (SA-MSG-10 park-and-freeze).
type DlqRecord struct {
	EventID      string
	Topic        string
	Key          string
	Payload      []byte
	Error        string
	AggregateKey string
}

// Dlq is the park-and-freeze adapter. A poison money/custody/inventory event is
// NEVER silently dropped: it is parked, and only its own AggregateKey freezes —
// other keys on the same topic keep progressing. Stateless; zero-value usable.
type Dlq struct{}

// Park records a poison event and freezes its aggregate key.
func (Dlq) Park(ctx context.Context, db DB, r DlqRecord) error {
	if db == nil {
		return fmt.Errorf("dlq: nil db handle")
	}
	if r.EventID == "" || r.AggregateKey == "" {
		return fmt.Errorf("dlq: eventId and aggregateKey are required")
	}
	_, err := db.ExecContext(ctx,
		`INSERT INTO dlq (event_id, topic, key, payload, error, aggregate_key, parked_at)
		 VALUES ($1,$2,$3,$4,$5,$6,now())`,
		r.EventID, r.Topic, r.Key, r.Payload, r.Error, r.AggregateKey)
	if err != nil {
		return fmt.Errorf("dlq: park %s: %w", r.EventID, err)
	}
	return nil
}

// IsKeyParked reports whether the given aggregate key is currently frozen. The
// consumer checks this before processing an event so a frozen key stays halted
// while sibling keys advance.
func (Dlq) IsKeyParked(ctx context.Context, db DB, aggregateKey string) (bool, error) {
	if db == nil {
		return false, fmt.Errorf("dlq: nil db handle")
	}
	rows, err := db.QueryContext(ctx,
		`SELECT 1 FROM dlq WHERE aggregate_key = $1 LIMIT 1`, aggregateKey)
	if err != nil {
		return false, fmt.Errorf("dlq: is-key-parked %s: %w", aggregateKey, err)
	}
	defer rows.Close()
	found := rows.Next()
	if err := rows.Err(); err != nil {
		return false, fmt.Errorf("dlq: is-key-parked %s: %w", aggregateKey, err)
	}
	return found, nil
}
